librad/src/profile.rs

```diff
@@ -0,0 +1,211 @@
+// Copyright © 2019-2021 The Radicle Foundation <hello@radicle.foundation>
+//
+// This file is part of radicle-link, distributed under the GPLv3 with Radicle
+// Linking Exception. For full terms see the included LICENSE file.
+
+use std::{
+    env,
+    fs,
+    io,
+    path,
+    path::{Path, PathBuf},
+};
+use uuid::Uuid;
+
+use crate::paths::{project_dirs, Paths};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("invalid profile ID in RAD_PROFILE environment variable: {id}")]
+    InvalidProfileIdFromEnv { id: String },
+    #[error("invalid profile ID loaded from {path}: {id}")]
+    InvalidProfileIdFromFile { id: String, path: PathBuf },
+    #[error(transparent)]
+    Io(#[from] io::Error),
+}
+
+/// A [`Profile`] provides [`Paths`] scoped by an identifier.
+///
+/// Profiles with different identifiers have distinct paths.
+#[derive(Debug, Clone)]
+pub struct Profile {
+    id: String,
+    paths: Paths,
+}
+
+impl Profile {
+    /// Creates a profile by loading the profile identifier and paths from
+    /// the environment variables or well-known file.
+    ///
+    /// If `RAD_PROFILE` is set, the value is used as the profile identifier
+    /// and system paths, scoped to the identifier, are used. See
+    /// [`Paths::new`] for details.
+    ///
+    /// Otherwise and if the `RAD_HOME` environment variable is set, the value
+    /// is used to construct the profile with [`Profile::from_root`].
+    ///
+    /// Otherwise, the profile identifier is read from a well-known file and
+    /// system paths scoped to the identifier are used. If the file containing
+    /// the identifier  does not exist, it is created and a new identifier is
+    /// generated and written to the file.
+    ///
+    /// The profile identifier must not empty, must not contain path separators,
+    /// must not be a windows path prefix like `C:`, and must not be a
+    /// special component like `.` or `..`.
+    ///
+    /// On Linux, the path to the active profile is
+    /// `$XDG_CONFIG_HOME/radicle-link/active_profile`.
+    pub fn load() -> Result<Self, Error> {
+        if let Ok(id) = env::var("RAD_PROFILE") {
+            if !is_valid_profile_id(&id) {
+                return Err(Error::InvalidProfileIdFromEnv { id });
+            }
+            let paths = Paths::new(&id)?;
+            return Ok(Self { id, paths });
+        }
+
+        if let Ok(rad_home) = env::var("RAD_HOME") {
+            return Self::from_root(Path::new(&rad_home));
```

Comment:
Shouldn't the `Paths` still contain the `profile id` here? Otherwise if you set `RAD_HOME` (eg when starting upstream) you wont have the multiple profile feature

Reply:
Good point. I was thinking about `RAD_HOME` as an exstension of profile selection where it would not only allow you to control the profile ID (like `RAD_PROFILE`) but also the location of all the data.

What you are suggesting would be to have `RAD_HOME` behave like setting all `XDG_*_HOME` variables all at once and having profiles work just like without `RAD_HOME` but under a different base directory.

I think this is the better approach as it decouples `RAD_HOME` and profiles and is a bit more predictable. @FintanH, @kim any thoughts on this?

Reply:
I'm not understanding the conversation here so I'll try and see how far my understanding goes.

So `RAD_HOME` decides where the root is which could contain multiple profiles and the active profile, i.e.
```
RAD_HOME
|--- active_profile
|--- 1
|    |--- git
|    |--- etc.
|--- 2
|    |---- git
```

But what @MeBrei is suggesting that `RAD_HOME` would point to _one_ of those profiles, for example, `RAD_HOME=<ROOT>/2`, is that correct?

Reply:
> So `RAD_HOME` decides where the root is which could contain multiple profiles and the active profile, i.e.
> 
> ```
> RAD_HOME
> |--- active_profile
> |--- 1
> |    |--- git
> |    |--- etc.
> |--- 2
> |    |---- git
> ```
> 
@FintanH the above is what I was suggesting. 
Below is what is currently happening, if you set `RAD_HOME` to a profile directory, but it in my tests I set it to a non-profile directory and it resulted in a mixed version, where the proxy store was under a profile id, but the librad directories weren't
> But what @MeBrei is suggesting that `RAD_HOME` would point to _one_ of those profiles, for example, `RAD_HOME=<ROOT>/2`, is that correct?

Reply:
Mmmm I think I'm piecing it together now. Something seems off to me which I commented here https://github.com/radicle-dev/radicle-link/pull/496/files#r566020184

Reply:
I’ve updated the PR (and added test) for to implement multi-profile support within `RAD_HOME`. I hope this should clarify things.

Reply:
I'm good with what's in c8e6f5d43c8ac8a94e5ba3039420a147e19a4a63.

Reply:
Sorry folks, didn’t actually push the updates. Could you have another look?

Reply:
Hm ok, so `RAD_HOME` and `RAD_PROFILE` are effectively mutually exclusive -- if I set `RAD_PROFILE`, I'd need to set `XDG_*_HOME` to make data go into a non-standard location (which is only possible under Linux). If I set `RAD_HOME`, the only way to control profile selection is via the `active_profile` file, which must already exist.

I find this _more_ confusing, but we all seem to have different expectations there. Perhaps we should continue to consider `RAD_HOME` as a convenience for test setups (not "stable API"), and look into respecting `XDG_*` variables regardless of platform?

Reply:
I also assumed both env variables could be combined. We already "released" `RAD_HOME` as a possibility to have an independent dev setup in upstream.

Reply:
As discussed in https://github.com/radicle-dev/radicle-link/issues/488#issuecomment-770923997 we will go with @MeBrei’s suggestion. This is implemented in f36bfc9a8aad7acdc637f5abd913a1c4dba61915